datafusion/core/src/physical_optimizer/limit_aggregation.rs

```diff
@@ -0,0 +1,88 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! An optimizer rule that detects aggregate operations that could use a limited bucket count
+
+use crate::physical_optimizer::PhysicalOptimizerRule;
+use crate::physical_plan::aggregates::AggregateExec;
+use crate::physical_plan::sorts::sort::SortExec;
+use crate::physical_plan::ExecutionPlan;
+use datafusion_common::config::ConfigOptions;
+use datafusion_common::{DataFusionError, Result};
+use std::sync::Arc;
+
+pub struct LimitAggregation {}
+
+impl LimitAggregation {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    fn recurse(plan: Arc<dyn ExecutionPlan>) -> Result<Arc<dyn ExecutionPlan>> {
```

Comment:
@alamb perhaps you can point me to the proper recursion helper here. It will need to find any AggregateExec with a SortExec above it in the tree where the sort is on the same `min/max` and there is a limit.

Also, it will need to check that any intermediate tree nodes don't mess with cardinality (i.e. outer joins)

Reply:
yeah, it doesn't like what I did: 
```
---- cli_quick_test::case_3_set_batch_size stdout ----
thread 'cli_quick_test::case_3_set_batch_size' panicked at /rustc/474709a9a2a74a8bcf0055fadb335d0ca0d2d939/library/core/src/ops/function.rs:250:5:
Unexpected stdout, failed var == "[{\"name\":\"datafusion.execution.batch_size\",\"setting\":\"1\"}]\n"
├── var: "limit aggregation\ncaused by\nInternal error: Children cannot be replaced in LazyMemTableExec { .. }. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker\n"
└── var as str: limit aggregation
    caused by
    Internal error: Children cannot be replaced in LazyMemTableExec { .. }. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```

Reply:
I think the standard DataFusion pattern is  add or extend one of the existing physical optimizer rules to recognize the specific pattern this operator handles 

Something like https://github.com/apache/arrow-datafusion/blob/main/datafusion/core/src/physical_optimizer/replace_with_order_preserving_variants.rs perhaps

The rewriting is handled by treenode/treenode rewriters: https://docs.rs/datafusion/latest/datafusion/common/tree_node/index.html

Reply:
Adding for reference the logical plan: 

```
Limit: skip=0, fetch=4
  Sort: MAX(traces.ts) DESC NULLS FIRST
    Projection: traces.trace_id, MAX(traces.ts)
      Aggregate: groupBy=[[traces.trace_id]], aggr=[[MAX(traces.ts)]]
        TableScan: traces
```
and physical plan:
```
GlobalLimitExec: skip=0, fetch=4
  SortExec: expr=[MAX(traces.ts)@1 DESC]
    AggregateExec: mode=Single, gby=[trace_id@0 as trace_id], aggr=[MAX(traces.ts)]
      CsvExec: file_groups={1 group: [[private/var/folders/fn/3l7phz3n14z5z5bg62y80cbc0000gp/T/.tmpsoXUWE/traces.csv]]}, projection=[trace_id, ts], has_header=false
```
from `test sql::group_by::group_by_limit`